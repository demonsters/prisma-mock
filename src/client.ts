import type { Prisma } from "@prisma/client"
import { createDelegate } from "./delegate"
import createIndexes from "./indexes"
import { MockPrismaOptions, PrismaMockData } from "./types"
import { deepCopy } from "./utils/deepCopy"
import { getCamelCase, removeMultiFieldIds } from "./utils/fieldHelpers"

// Creates a mock Prisma client.
// @param prisma - The Prisma namespace or client constructor.
// @param options - Options for configuring the mock client:
//   - data: Initial mock data for your models (default: {}).
//   - caseInsensitive: If true, string matching is case-insensitive (default: false).
//   - enableIndexes: If true, enables index lookups for performance (default: false).
//   - mockClient: Optionally provide your own mock client (jest-mock-extended or vitest-mock-extended) instance to use.
// @returns A mock Prisma client with all model methods and access to internal state.
const createPrismaMock = <P>(
  prisma: typeof Prisma,
  options: MockPrismaOptions = {
    caseInsensitive: false,
    enableIndexes: false,
    data: {}
  }
): P & {
  $getInternalState: () => Required<PrismaMockData<P>>
} => {

  // Reference object to hold the mock data state
  let ref = {
    data: options.data || {},
  }

  // Initialize the mock client (either use provided one or create new)
  let client = options.mockClient ? options.mockClient : {}

  /**
   * Helper function to implement mock methods consistently
   * @param name - Method name to mock
   * @param fnc - Function implementation
   */
  const mockImplementation = (name: string, fnc: any) => {
    if (options.mockClient) {
      client[name].mockImplementation(fnc)
    } else {
      client[name] = fnc
    }
  }

  // Create indexes if enabled in options
  const indexes = createIndexes(!!options.enableIndexes, prisma)

  // Determine if case-insensitive matching should be used
  const caseInsensitive = options.caseInsensitive || false

  // Mock $transaction method for handling database transactions
  mockImplementation("$transaction", async (actions: Promise<any>[] | ((prisma: P) => Promise<any>)) => {
    const res = []
    if (Array.isArray(actions)) {
      // Handle array of promises (parallel execution)
      for (const action of actions) {
        res.push(await action)
      }
      return res
    } else {
      // Handle callback function (serial execution with rollback on error)
      const snapshot = deepCopy(ref.data)
      try {
        // @ts-ignore
        return await actions(client)
      } catch (error) {
        // Rollback data on error
        ref.data = snapshot
        throw error
      }
    }
  })

  // Mock connection methods
  mockImplementation("$connect", async () => { })
  mockImplementation("$disconnect", async () => { })
  mockImplementation("$use", async () => {
    throw new Error("$use is not yet implemented in prisma-mock")
  })

  // Create delegate functions for model operations
  const Delegate = createDelegate(ref, prisma, caseInsensitive, indexes)

  // Initialize each model in the datamodel
  prisma.dmmf.datamodel.models.forEach((model) => {
    if (!model) return

    // Convert model name to camelCase for consistency
    const c = getCamelCase(model.name)

    // Initialize empty array for model if it doesn't exist
    if (!ref.data[c]) {
      ref.data = {
        ...(ref.data || {}),
        [c]: [],
      }
    }

    // Remove multi-field IDs from the data structure
    ref.data = removeMultiFieldIds(model, ref.data)

    // Set up indexes for each field in the model
    model.fields.forEach((field) => {
      const isPrimaryKey = !!model.primaryKey?.fields.includes(field.name)
      indexes.addIndexFieldIfNeeded(c, field, isPrimaryKey)
    })

    // Update indexes with existing data
    ref.data[c].forEach((item) => {
      indexes.updateItem(c, item, null)
    })

    // Create delegate functions for this model
    const objs = Delegate(c, model)

    // Bind delegate functions to the client
    Object.keys(objs).forEach((fncName) => {
      // Skip private methods (those starting with underscore)
      if (fncName.indexOf("_") === 0) return

      // Initialize model namespace if it doesn't exist
      if (!client[c]) client[c] = {}

      // Bind the delegate function to the client
      if (options.mockClient) {
        client[c][fncName].mockImplementation(async (...params) => {
          return objs[fncName](...params)
        })
      } else {
        client[c][fncName] = async (...params) => {
          return objs[fncName](...params)
        }
      }
    })
  })

  // Add method to access internal state for testing/debugging
  client['$getInternalState'] = () => ref.data

  // @ts-ignore
  return client
}

export default createPrismaMock
