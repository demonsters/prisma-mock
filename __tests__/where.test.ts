// @ts-nocheck

import createPrismaClient from "./createPrismaClient"
import { Role } from "@prisma/client"

describe("PrismaClient where", () => {
  const date1 = new Date('2020-02-01T14:00:00.000Z')
  const date2 = new Date('2020-02-02T14:00:00.000Z')

  const data = {
    account: [
      {
        name: "Piet",
      },
      {
        name: "Dirk",
      },
      {
        name: "Sjors",
      },
      {
        name: "ABBY",
      },
      {
        name: null
      }
    ],
    user: [
      {
        id: 1,
        name: "Henk",
        accountId: 1,
        uniqueField: "user 1",
      },
      {
        id: 2,
        name: "Dirk",
        accountId: 2,
        uniqueField: "user 2",
      },
    ],
    post: [
      {
        title: "Piet",
        updated: date1,
        created: date1,
      },
      {
        title: "Dirk",
        updated: date2,
        created: date2,
      },
      {
        title: "Sjors",
        updated: date2,
        created: date2,
      },
      {
        title: "ABBY",
        updated: date2,
        created: date2,
      },
    ],
    element: [
      {
        e_id: 1,
        json: {
          data: "first"
        },
        userId: 1,
        value: "value 1",
      },
      {
        e_id: 2,
        json: {
          data: "second"
        },
        userId: 2,
        value: "value 2",
      }
    ],
    stripe: [
      {
        id: 1,
        customerId: "cus_123",
        accountId: 2,
      },
    ],
  }

  describe.each([false, true])(
    "caseInsensitive %p",
    (caseInsensitive: boolean) => {

      describe("No mode", () => {

        if (process.env.PROVIDER === "postgresql" && caseInsensitive) {
          // Can't test case insensitive for postgresql without mode right now
          return
        }

        test("multiple", async () => {
          const client = await createPrismaClient(data, {
            caseInsensitive,
          })
          const accounts = await client.post.findMany({
            where: {
              created: { lte: date1 },
              title: caseInsensitive ? "dirk" : "Dirk",
            },
          })
          expect(accounts.length).toBe(0)
          expect(accounts).toMatchSnapshot()
          // expect(account).not.toEqual([data.post[0]]);
        })

        test("notIn", async () => {
          const client = await createPrismaClient(data, {
            caseInsensitive,
          })
          const accounts = await client.account.findMany({
            where: {
              name: { notIn: ["Piet", "Sjors"] },
            },
          })
          expect(accounts).toMatchSnapshot()
          expect(accounts.length).toEqual(2)
          // expect(account).toEqual([data.account[1], data.account[3]]);
        })

        test("Json: equals", async () => {
          const client = await createPrismaClient(data, {
            caseInsensitive,
          })
          const elements = await client.element.findMany({
            where: {
              json: {
                equals: {
                  data: "first"
                }
              }
            },
          })
          expect(elements).toMatchSnapshot()
          // expect(element).toEqual([data.element[0]]);
        })
      
        test("in", async () => {
          const client = await createPrismaClient(data, {
            caseInsensitive,
          })
          const accounts = await client.account.findMany({
            where: {
              name: { in: ["Piet", "Sjors"] },
            },
          })
          expect(accounts).toMatchSnapshot()
          expect(accounts.length).toEqual(2)
          // expect(account).toEqual([data.account[0], data.account[2]]);
        })
      
        test("in (nested)", async () => {
          const client = await createPrismaClient(data, {
            caseInsensitive,
          })
          const accounts = await client.user.findMany({
            where: {
              account: {
                name: { in: ["Piet", "Sjors"] },
              }
            },
          })
          expect(accounts).toMatchSnapshot()
          expect(accounts.length).toEqual(1)
          // expect(account).toEqual([data.account[0], data.account[2]]);
        })

        test("Nested", async () => {
          const client = await createPrismaClient(data, {
            caseInsensitive,
          })
          const accounts = await client.user.findMany({
            where: {
              account: {
                name: "Dirk",
              },
            },
          })
          expect(accounts.length).toEqual(1)
          expect(accounts).toMatchSnapshot()
        })

        test("OR", async () => {
          const client = await createPrismaClient(data, {
            caseInsensitive,
          })
          const accounts = await client.account.findMany({
            where: {
              OR: [{ name: "Dirk" }, { name: "Piet" }],
            },
          })
          expect(accounts.length).toEqual(2)
          expect(accounts).toMatchSnapshot()
          // expect(accounts).toEqual([data.account[0], data.account[1]]);
        })

        test("NOT array", async () => {
          const client = await createPrismaClient(data, {
            caseInsensitive,
          })
          const accounts = await client.account.findMany({
            where: {
              NOT: [{ name: "Dirk" }, { name: "Piet" }],
            },
          })
          // expect(accounts.length).toEqual(3)
          expect(accounts).toMatchSnapshot()
          // expect(accounts).toEqual([data.account[2], data.account[3]]);
        })

        test("NOT object", async () => {
          const client = await createPrismaClient(data, {
            caseInsensitive,
          })
          const accounts = await client.account.findMany({
            where: {
              NOT: { name: "Dirk" },
            },
          })
          expect(accounts.length).toEqual(3)
          expect(accounts).toMatchSnapshot()
          // expect(accounts).toEqual([data.account[2], data.account[3]]);
        })

        test("AND", async () => {
          const client = await createPrismaClient(data, {
            caseInsensitive,
          })
          const accounts = await client.account.findMany({
            where: {
              AND: [{ name: "Dirk" }, { id: 2 }],
            },
          })
          expect(accounts.length).toEqual(1)
          expect(accounts).toMatchSnapshot()
          // expect(accounts).toEqual([data.account[1]]);
        })

        test("AND (failing)", async () => {
          const client = await createPrismaClient(data, {
            caseInsensitive,
          })
          const accounts = await client.account.findMany({
            where: {
              AND: [{ id: 1 }, { id: 2 }],
            },
          })
          expect(accounts.length).toEqual(0)
          expect(accounts).toMatchSnapshot()
          // expect(accounts).toEqual([data.account[1]]);
        })
      })

      describe.each([false, true])(
        "mode %p",
        (mode: boolean) => {

          if (process.env.PROVIDER === "postgresql" && !mode && caseInsensitive) {
            // Can't test case insensitive for postgresql without mode right now
            return
          }

          let client: PrismaClient
          beforeEach(async () => {
            client = await createPrismaClient(data, mode ? undefined : {
              caseInsensitive: mode ? false : caseInsensitive,
            })
          })

          test("startsWith", async () => {
            const accounts = await client.account.findMany({
              where: {
                name: {
                  startsWith: caseInsensitive ? "di" : "Di",
                  mode: caseInsensitive && mode ? "insensitive" : "default",
                },
              },
            })
            expect(accounts.length).toBe(1)
            expect(accounts).toMatchSnapshot()
            // expect(account).toEqual([data.account[1]]);
          })

          test("endsWith", async () => {
            const client = await createPrismaClient(data, mode ? undefined : {
              caseInsensitive: mode ? false : caseInsensitive,
            })
            const accounts = await client.account.findMany({
              where: {
                name: {
                  endsWith: "rk",
                  mode: caseInsensitive && mode ? "insensitive" : "default",
                },
              },
            })
            expect(accounts).toMatchSnapshot()
            // expect(account).toEqual([data.account[1]]);
          })

          
          test("contains", async () => {
            const accounts = await client.account.findMany({
              where: {
                name: {
                  contains: "BB",
                  mode: caseInsensitive && mode ? "insensitive" : "default",
                },
              },
            })
            expect(accounts).toMatchSnapshot()
            // expect(account).toEqual([data.account[3]]);
          })

          test("equals", async () => {
            const accounts = await client.account.findMany({
              where: {
                name: {
                  equals: caseInsensitive ? "dirk" : "Dirk",
                  mode: caseInsensitive && mode ? "insensitive" : "default",
                },
              },
            })
            expect(accounts.length).toEqual(1)
            expect(accounts).toMatchSnapshot()
            // expect(account).toEqual([data.account[1]]);
          })
        }
      )
    }
  )

  describe("Case inrelevant", () => {

    test("gt", async () => {
      const client = await createPrismaClient(data)
      const accounts = await client.account.findMany({
        where: {
          id: { gt: 1 },
        },
      })
      expect(accounts.length).toEqual(4)
      expect(accounts).toMatchSnapshot()
      // expect(account).toEqual([
      //   data.account[1],
      //   data.account[2],
      //   data.account[3],
      // ]);
    })

    test("gte", async () => {
      const client = await createPrismaClient(data)
      const accounts = await client.account.findMany({
        where: {
          id: { gte: 1 },
        },
      })
      expect(accounts).toMatchSnapshot()
      // expect(account).toEqual(data.account);
    })

    test("lt", async () => {
      const client = await createPrismaClient(data)
      const accounts = await client.account.findMany({
        where: {
          id: { lt: 2 },
        },
      })
      expect(accounts).toMatchSnapshot()
      expect(accounts.length).toEqual(1)
      // expect(account).toEqual([data.account[0]]);
    })

    test("lte", async () => {
      const client = await createPrismaClient(data)
      const accounts = await client.account.findMany({
        where: {
          id: { lte: 2 },
        },
      })
      expect(accounts).toMatchSnapshot()
      expect(accounts.length).toEqual(2)
      // expect(account).toEqual([data.account[0], data.account[1]]);
    })

    test("not", async () => {
      const client = await createPrismaClient(data)
      const accounts = await client.account.findMany({
        where: {
          id: { not: 2 },
        },
      })
      expect(accounts.length).toEqual(4)
      expect(accounts).toMatchSnapshot()
      // expect(account).toEqual([
      //   data.account[0],
      //   data.account[2],
      //   data.account[3],
      // ]);
    })

    test("Json: not", async () => {
      const client = await createPrismaClient(data)
      const elements = await client.element.findMany({
        where: {
          json: {
            not: {
              data: "first"
            }
          }
        },
      })
      expect(elements).toMatchSnapshot()
      expect(elements.length).toEqual(1)
      // expect(elements).toEqual([data.element[1]]);
    })

    test("Nested deep", async () => {
      const client = await createPrismaClient(data)
      const accounts = await client.user.findMany({
        where: {
          account: {
            stripe: {
              customerId: "cus_123",
            },
          },
        },
      })
      expect(accounts.length).toEqual(1)
      expect(accounts).toMatchSnapshot()
    })

    test("date", async () => {
      const client = await createPrismaClient(data)
      // Clone date & time
      const accounts = await client.post.findMany({
        where: {
          created: new Date(date1.getTime()),
        },
      })
      // expect(accounts).toEqual([data.post[0]]);
      expect(accounts.length).toEqual(1)
      expect(accounts).toMatchSnapshot()
    })
  })

  describe("join", () => {
    const data = {
      account: [
        { id: 1, name: "A" },
        { id: 2, name: "B" },
        { id: 3, name: "C" },
      ],
      user: [
        { id: 1, accountId: 1, role: Role.ADMIN, uniqueField: "user 1" },
        { id: 2, accountId: 1, role: Role.ADMIN, uniqueField: "user 2" },
        { id: 3, accountId: 2, role: Role.USER, uniqueField: "user 3" },
        { id: 4, accountId: 2, role: Role.ADMIN, uniqueField: "user 4" },
      ],
    }

    test("every", async () => {
      const client = await createPrismaClient(data)
      const accounts = await client.account.findMany({
        where: {
          users: {
            every: {
              role: Role.ADMIN,
            },
          },
        },
      })
      expect(accounts).toMatchInlineSnapshot(`
Array [
  Object {
    "id": 1,
    "name": "A",
    "sort": null,
  },
  Object {
    "id": 3,
    "name": "C",
    "sort": null,
  },
]
`)
    })
    test("some", async () => {
      const client = await createPrismaClient(data)
      const accounts = await client.account.findMany({
        where: {
          users: {
            some: {
              role: Role.ADMIN,
            },
          },
        },
      })
      expect(accounts).toMatchInlineSnapshot(`
Array [
  Object {
    "id": 1,
    "name": "A",
    "sort": null,
  },
  Object {
    "id": 2,
    "name": "B",
    "sort": null,
  },
]
`)
      // expect(accounts.length).toEqual(2);
      // expect(accounts).toEqual([data.account[0], data.account[1]]);
    })

    test("none", async () => {
      const client = await createPrismaClient(data)
      const accounts = await client.account.findMany({
        where: {
          users: {
            none: {
              role: Role.ADMIN,
            },
          },
        },
      })
      expect(accounts).toMatchInlineSnapshot(`
Array [
  Object {
    "id": 3,
    "name": "C",
    "sort": null,
  },
]
`)
      // expect(accounts.length).toEqual(1);
      // expect(accounts).toEqual([data.account[2]]);
    })

    // every?: PostWhereInput | null
    // some?: PostWhereInput | null
    // none?: PostWhereInput | null
  })

  describe("null & undefined", () => {
    const data = {
      user: [
        {
          id: 1,
          name: "Henk",
          uniqueField: "user 1",
        },
        {
          id: 2,
          name: undefined,
          uniqueField: "user 2",
        },
      ],
    }

    test("null", async () => {
      const client = await createPrismaClient(data)
      const users = await client.user.findMany({
        where: {
          name: null,
        },
      })
      expect(users.length).toEqual(1)
    })

    test("undefined", async () => {
      const client = await createPrismaClient(data)
      const users = await client.user.findMany({
        where: {
          name: undefined,
        },
      })
      expect(users.length).toEqual(2)
    })
  })
})
