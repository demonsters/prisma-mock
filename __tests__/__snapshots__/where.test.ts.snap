// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`PrismaClient where Case inrelevant Json: not 1`] = `
[
  {
    "e_id": 2,
    "json": {
      "data": "second",
    },
    "userId": 2,
    "value": "value 2",
  },
]
`;

exports[`PrismaClient where Case inrelevant Nested deep 1`] = `
[
  {
    "accountId": 2,
    "age": 10,
    "clicks": null,
    "deleted": false,
    "id": 2,
    "name": "Dirk",
    "role": "ADMIN",
    "sort": null,
    "uniqueField": "user 2",
  },
]
`;

exports[`PrismaClient where Case inrelevant date 1`] = `
[
  {
    "authorId": null,
    "created": 2020-02-01T14:00:00.000Z,
    "id": 1,
    "published": false,
    "title": "Piet",
    "updated": 2020-02-01T14:00:00.000Z,
  },
]
`;

exports[`PrismaClient where Case inrelevant gt 1`] = `
[
  {
    "id": 2,
    "name": "Dirk",
    "sort": null,
  },
  {
    "id": 3,
    "name": "Sjors",
    "sort": null,
  },
  {
    "id": 4,
    "name": "ABBY",
    "sort": null,
  },
  {
    "id": 5,
    "name": null,
    "sort": null,
  },
]
`;

exports[`PrismaClient where Case inrelevant gte 1`] = `
[
  {
    "id": 1,
    "name": "Piet",
    "sort": null,
  },
  {
    "id": 2,
    "name": "Dirk",
    "sort": null,
  },
  {
    "id": 3,
    "name": "Sjors",
    "sort": null,
  },
  {
    "id": 4,
    "name": "ABBY",
    "sort": null,
  },
  {
    "id": 5,
    "name": null,
    "sort": null,
  },
]
`;

exports[`PrismaClient where Case inrelevant lt 1`] = `
[
  {
    "id": 1,
    "name": "Piet",
    "sort": null,
  },
]
`;

exports[`PrismaClient where Case inrelevant lte 1`] = `
[
  {
    "id": 1,
    "name": "Piet",
    "sort": null,
  },
  {
    "id": 2,
    "name": "Dirk",
    "sort": null,
  },
]
`;

exports[`PrismaClient where Case inrelevant not 1`] = `
[
  {
    "id": 1,
    "name": "Piet",
    "sort": null,
  },
  {
    "id": 3,
    "name": "Sjors",
    "sort": null,
  },
  {
    "id": 4,
    "name": "ABBY",
    "sort": null,
  },
  {
    "id": 5,
    "name": null,
    "sort": null,
  },
]
`;

exports[`PrismaClient where caseInsensitive false No mode AND (failing) 1`] = `[]`;

exports[`PrismaClient where caseInsensitive false No mode AND 1`] = `
[
  {
    "id": 2,
    "name": "Dirk",
    "sort": null,
  },
]
`;

exports[`PrismaClient where caseInsensitive false No mode Json: equals 1`] = `
[
  {
    "e_id": 1,
    "json": {
      "data": "first",
    },
    "userId": 1,
    "value": "value 1",
  },
]
`;

exports[`PrismaClient where caseInsensitive false No mode NOT array 1`] = `
[
  {
    "id": 3,
    "name": "Sjors",
    "sort": null,
  },
  {
    "id": 4,
    "name": "ABBY",
    "sort": null,
  },
]
`;

exports[`PrismaClient where caseInsensitive false No mode NOT object 1`] = `
[
  {
    "id": 1,
    "name": "Piet",
    "sort": null,
  },
  {
    "id": 3,
    "name": "Sjors",
    "sort": null,
  },
  {
    "id": 4,
    "name": "ABBY",
    "sort": null,
  },
]
`;

exports[`PrismaClient where caseInsensitive false No mode Nested 1`] = `
[
  {
    "accountId": 2,
    "age": 10,
    "clicks": null,
    "deleted": false,
    "id": 2,
    "name": "Dirk",
    "role": "ADMIN",
    "sort": null,
    "uniqueField": "user 2",
  },
]
`;

exports[`PrismaClient where caseInsensitive false No mode OR 1`] = `
[
  {
    "id": 1,
    "name": "Piet",
    "sort": null,
  },
  {
    "id": 2,
    "name": "Dirk",
    "sort": null,
  },
]
`;

exports[`PrismaClient where caseInsensitive false No mode in (nested) 1`] = `
[
  {
    "accountId": 1,
    "age": 10,
    "clicks": null,
    "deleted": false,
    "id": 1,
    "name": "Henk",
    "role": "ADMIN",
    "sort": null,
    "uniqueField": "user 1",
  },
]
`;

exports[`PrismaClient where caseInsensitive false No mode in 1`] = `
[
  {
    "id": 1,
    "name": "Piet",
    "sort": null,
  },
  {
    "id": 3,
    "name": "Sjors",
    "sort": null,
  },
]
`;

exports[`PrismaClient where caseInsensitive false No mode multiple 1`] = `[]`;

exports[`PrismaClient where caseInsensitive false No mode notIn 1`] = `
[
  {
    "id": 2,
    "name": "Dirk",
    "sort": null,
  },
  {
    "id": 4,
    "name": "ABBY",
    "sort": null,
  },
]
`;

exports[`PrismaClient where caseInsensitive false mode false contains 1`] = `
[
  {
    "id": 4,
    "name": "ABBY",
    "sort": null,
  },
]
`;

exports[`PrismaClient where caseInsensitive false mode false endsWith 1`] = `
[
  {
    "id": 2,
    "name": "Dirk",
    "sort": null,
  },
]
`;

exports[`PrismaClient where caseInsensitive false mode false equals 1`] = `
[
  {
    "id": 2,
    "name": "Dirk",
    "sort": null,
  },
]
`;

exports[`PrismaClient where caseInsensitive false mode false startsWith 1`] = `
[
  {
    "id": 2,
    "name": "Dirk",
    "sort": null,
  },
]
`;

exports[`PrismaClient where caseInsensitive false mode true contains 1`] = `
[
  {
    "id": 4,
    "name": "ABBY",
    "sort": null,
  },
]
`;

exports[`PrismaClient where caseInsensitive false mode true endsWith 1`] = `
[
  {
    "id": 2,
    "name": "Dirk",
    "sort": null,
  },
]
`;

exports[`PrismaClient where caseInsensitive false mode true equals 1`] = `
[
  {
    "id": 2,
    "name": "Dirk",
    "sort": null,
  },
]
`;

exports[`PrismaClient where caseInsensitive false mode true startsWith 1`] = `
[
  {
    "id": 2,
    "name": "Dirk",
    "sort": null,
  },
]
`;

exports[`PrismaClient where caseInsensitive true No mode AND (failing) 1`] = `[]`;

exports[`PrismaClient where caseInsensitive true No mode AND 1`] = `
[
  {
    "id": 2,
    "name": "Dirk",
    "sort": null,
  },
]
`;

exports[`PrismaClient where caseInsensitive true No mode Json: equals 1`] = `
[
  {
    "e_id": 1,
    "json": {
      "data": "first",
    },
    "userId": 1,
    "value": "value 1",
  },
]
`;

exports[`PrismaClient where caseInsensitive true No mode NOT array 1`] = `
[
  {
    "id": 3,
    "name": "Sjors",
    "sort": null,
  },
  {
    "id": 4,
    "name": "ABBY",
    "sort": null,
  },
  {
    "id": 5,
    "name": null,
    "sort": null,
  },
]
`;

exports[`PrismaClient where caseInsensitive true No mode NOT object 1`] = `
[
  {
    "id": 1,
    "name": "Piet",
    "sort": null,
  },
  {
    "id": 3,
    "name": "Sjors",
    "sort": null,
  },
  {
    "id": 4,
    "name": "ABBY",
    "sort": null,
  },
]
`;

exports[`PrismaClient where caseInsensitive true No mode Nested 1`] = `
[
  {
    "accountId": 2,
    "age": 10,
    "clicks": null,
    "deleted": false,
    "id": 2,
    "name": "Dirk",
    "role": "ADMIN",
    "sort": null,
    "uniqueField": "user 2",
  },
]
`;

exports[`PrismaClient where caseInsensitive true No mode OR 1`] = `
[
  {
    "id": 1,
    "name": "Piet",
    "sort": null,
  },
  {
    "id": 2,
    "name": "Dirk",
    "sort": null,
  },
]
`;

exports[`PrismaClient where caseInsensitive true No mode in (nested) 1`] = `
[
  {
    "accountId": 1,
    "age": 10,
    "clicks": null,
    "deleted": false,
    "id": 1,
    "name": "Henk",
    "role": "ADMIN",
    "sort": null,
    "uniqueField": "user 1",
  },
]
`;

exports[`PrismaClient where caseInsensitive true No mode in 1`] = `
[
  {
    "id": 1,
    "name": "Piet",
    "sort": null,
  },
  {
    "id": 3,
    "name": "Sjors",
    "sort": null,
  },
]
`;

exports[`PrismaClient where caseInsensitive true No mode multiple 1`] = `[]`;

exports[`PrismaClient where caseInsensitive true No mode notIn 1`] = `
[
  {
    "id": 2,
    "name": "Dirk",
    "sort": null,
  },
  {
    "id": 4,
    "name": "ABBY",
    "sort": null,
  },
]
`;

exports[`PrismaClient where caseInsensitive true mode false contains 1`] = `
[
  {
    "id": 4,
    "name": "ABBY",
    "sort": null,
  },
]
`;

exports[`PrismaClient where caseInsensitive true mode false endsWith 1`] = `
[
  {
    "id": 2,
    "name": "Dirk",
    "sort": null,
  },
]
`;

exports[`PrismaClient where caseInsensitive true mode false equals 1`] = `
[
  {
    "id": 2,
    "name": "Dirk",
    "sort": null,
  },
]
`;

exports[`PrismaClient where caseInsensitive true mode false startsWith 1`] = `
[
  {
    "id": 2,
    "name": "Dirk",
    "sort": null,
  },
]
`;

exports[`PrismaClient where caseInsensitive true mode true contains 1`] = `
[
  {
    "id": 4,
    "name": "ABBY",
    "sort": null,
  },
]
`;

exports[`PrismaClient where caseInsensitive true mode true endsWith 1`] = `
[
  {
    "id": 2,
    "name": "Dirk",
    "sort": null,
  },
]
`;

exports[`PrismaClient where caseInsensitive true mode true equals 1`] = `
[
  {
    "id": 2,
    "name": "Dirk",
    "sort": null,
  },
]
`;

exports[`PrismaClient where caseInsensitive true mode true startsWith 1`] = `
[
  {
    "id": 2,
    "name": "Dirk",
    "sort": null,
  },
]
`;
